### ISSUE TITLE
Atomic Operation Tests fail

### ISSUE THREAD

**Issue by harryherold** [cc-hpc-itwm/GPI-2#31]
We installed GPI2-1.3.0 on our cluster.
All tests are successful on the cluster partition with the Intel Sandy architecture.
But when I execute the tests on the partition with the Intel Haswell architecture, 
the atomic operation tests fail with this output:

> atomic operations are not supported yet
> Assertion failed in fetch_add.c[25]: Return -1: general error

On the cluster is installed:
- gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-16)
- libibverbs-1.1.8-1.ofed3.18.Bull.1.0.el6.20150922.x86_64

Have you some advises or experience with this error?

**Comment by rumach** [comment]
Hi Christian,
is there a description of the system (hw) somewhere online?

cheers.

**Comment by harryherold** [comment]
Yes, look here

https://tu-dresden.de/die_tu_dresden/zentrale_einrichtungen/zih/hpc/hochleistungsrechner

There you find the regarding system in the section called  "Phase 2".
Additionally, here is a more general description:

https://doc.zih.tu-dresden.de/hpc-wiki/bin/view/Compendium/SystemTaurus

Let me know If you need any other things.

**Comment by rumach** [comment]
Thanks for the info. Maybe I just rushed but I couldn't find informations on the network. You referred the libibverbs so I suppose it's Infiniband. But what kind of card?
cheers.

**Comment by harryherold** [comment]
I got this information from lspci:

> ...
> 04:00.0 Infiniband controller: Mellanox Technologies MT27600 [Connect-IB]
> ...

Here is the output from ibv_devinfo:

hca_id:	mlx5_0
	transport:			InfiniBand (0)
	fw_ver:				10.12.1100
	node_guid:			0800:3800:013c:4b37
	sys_image_guid:			0800:3800:013c:4b37
	vendor_id:			0x119f
	vendor_part_id:			7059
	hw_ver:				0x0
	board_id:			BL_12001158CIBF1
	phys_port_cnt:			1
		port:	1
			state:			PORT_ACTIVE (4)
			max_mtu:		4096 (5)
			active_mtu:		4096 (5)
			sm_lid:			28
			port_lid:		1522
			port_lmc:		0x00
			link_layer:		InfiniBand

**Comment by rumach** [comment]
OK, thanks.
Is there any more information or just those 2 messages (e.g. when you run the test directly with gaspi_run)?
The reason I ask is that I don't recall and can't find the message "atomic operations are not supported yet". It must be coming from somewhere else.
Does the system have the utility ib_atomic_lat? If yes, can you run it between two nodes?

**Comment by harryherold** [comment]
Message comes from here:

http://www.spinics.net/lists/linux-rdma/msg31162.html

**Comment by harryherold** [comment]
ib_atomic_lat - server shows:
Device not recognized to implement inline feature. Disabling it
---------------------------------------------------------------------------------------

************************************
* Waiting for client to connect... *
************************************
---------------------------------------------------------------------------------------
                    Atomic FETCH_AND_ADD Latency Test
 Dual-port       : OFF		Device         : mlx5_0
 Number of qps   : 1		Transport type : IB
 Connection type : RC		Using SRQ      : OFF
 Mtu             : 4096[B]
 Link type       : IB
 Outstand reads  : 16
 rdma_cm QPs	 : OFF
 Data ex. method : Ethernet
---------------------------------------------------------------------------------------
 local address: LID 0x20e QPN 0x447c PSN 0x396694
 remote address: LID 0x132 QPN 0x25d4 PSN 0xe3523e
ethernet_read_keys: Couldn't read remote address
 Unable to read to socket/rdam_cm
Failed to close connection between server and client
srun: error: taurusi6558: task 0: Exited with exit code 1

And the client shows:

**Comment by rumach** [comment]
OK, I see now where that message comes from (I haven't seen it before). Do you really need the support for atomics?

**Comment by harryherold** [comment]
Yes, we need this for a thesis.

**Comment by rumach** [comment]
I see. Unfortunately I think it is something outside of GPI-2, at least as long as the Infiniband utility reports an error.

**Comment by harryherold** [comment]
So, we patched the libmlx5 and the ib_atomic_lat runs without failure.
Unfortunately, the fetch-and-add test fails with this message:

> Assertion failed in fetch_add.c[25]: Return -1: general error

Have you a Haswell system where GPI2 is running?

**Comment by rumach** [comment]
Do you mind trying with the branch next (with some extra):
GPI2_EXTRA_CFLAGS=-DGPI2_EXP_VERBS sh -x ./install.sh -p <path_to_installation>

and let me know if your problem still appears? Thanks.

**Comment by Flamefire** [comment]
I just run into the same issue. Some debugging revealed that on Sandy Bridge libmlx4 is used where atomics are implemented. On Haswell libmlx5 is used where it isn't.

There is some more discussion about this here: https://github.com/open-mpi/ompi/issues/1077
TLDR:

> The problem is that with mlx5 the atomic result is returned in network byte order. Because of this they disabled the atomics when using ibv_post_send. It is possible to use the atomics if verbs extensions are installed and ibv_exp_post_send is used

They have a patch here: https://github.com/open-mpi/ompi/pull/1147

I tried to backport your commits using `ibv_exp_post_send` but unfortunately the verbs extension is not installed.

**Comment by Flamefire** [comment]
And finally: The Open Fabrics Alliance OFED stack should support atomics from version 4.8+, not the 3.x. The Verbs Extensions are only available for the Mellanox OFED stack, so using them is not a solution.
